libs/ledger/src/protocols/notarisation_service.cpp

```diff
@@ -0,0 +1,399 @@
+
+      FETCH_LOG_INFO(LOGGING_NAME, "Notarisation built for block ", hash.ToHex());
+      assert(notarisations_built_[notarisation_collection_height_].find(hash) ==
+             notarisations_built_[notarisation_collection_height_].end());
+      notarisations_built_[notarisation_collection_height_][hash] = sig;
+    }
+  }
+
+  return State::COMPLETE;
+}
+
+NotarisationService::State NotarisationService::OnComplete()
+{
+  FETCH_LOCK(mutex_);
+
+  // TODO(JMW): Clear old signature shares
+
+  // If chain has moved ahead faster while notarisations were being collected then reset
+  // the collection height
+  if (notarised_chain_height_ + 1 > notarisation_collection_height_)
+  {
+    notarisation_collection_height_ = notarised_chain_height_ + 1;
+  }
+
+  // Completed notarisation of a sequence of blocks during aeon. Any notarised blocks not received
+  // through RPC will be obtained via broadcast
+  if (new_keys)
+  {
+    // Save previous as we might still need to notarise blocks from previous aeon
+    previous_notarisation_unit_ = active_notarisation_unit_;
+    active_notarisation_unit_.reset();
+    return State::KEY_ROTATION;
+  }
+
+  return State::COLLECT_NOTARISATIONS;
+}
+
+NotarisationService::BlockNotarisationShares NotarisationService::GetNotarisations(
+    BlockHeight const &height)
+{
+  FETCH_LOCK(mutex_);
+  if (notarisations_being_built_.find(height) == notarisations_being_built_.end())
+  {
+    return {};
+  }
+  return notarisations_being_built_.at(height);
+}
+
+void NotarisationService::NotariseBlock(BlockBody const &block)
+{
+  FETCH_LOCK(mutex_);
+
+  // Return if block is too far in the past of the head of the notarised chain
+  if (block.block_number < BlockNumberCutoff())
+  {
+    return;
+  }
+  // Return if block has already been notarised
+  if (notarisations_built_.find(block.block_number) != notarisations_built_.end() &&
+      notarisations_built_.at(block.block_number).find(block.hash) !=
+          notarisations_built_.at(block.block_number).end())
+  {
+    return;
+  }
+
+  // If block is within current aeon then notarise now
+  SharedAeonNotarisationUnit notarisation_unit;
+  if (active_notarisation_unit_ && block.block_number >= active_notarisation_unit_->round_start() &&
+      block.block_number < active_notarisation_unit_->round_end())
+  {
+    notarisation_unit = active_notarisation_unit_;
+  }
+
+  // If block is not in current aeon then check previous notarisation unit
+  if (previous_notarisation_unit_ &&
+      block.block_number >= previous_notarisation_unit_->round_start() &&
+      block.block_number < previous_notarisation_unit_->round_end())
+  {
+    notarisation_unit = previous_notarisation_unit_;
+  }
+
+  if (!notarisation_unit)
+  {
+    return;
+  }
+
+  // Determine rank of miner in qual
+  auto entropy_ranked_cabinet = Consensus::QualWeightedByEntropy(
+      notarisation_unit->notarisation_members(), block.block_entropy.EntropyAsU64());
```

Comment:
From previous review comment regarding having the block coordinator only send the notarisation service ever higher ranked blocks: I left this here as I don't think the block coordinator should keep track of what ranked blocks the notarisation service has notarised.